tools/setup_helpers/extension.py

```diff
@@ -0,0 +1,106 @@
+import os
+import sys
+import platform
+import subprocess
+from pathlib import Path
+
+import torch
+from setuptools import Extension
+from setuptools.command.build_ext import build_ext
+
+__all__ = [
+    "get_ext_modules",
+    "CMakeBuild",
+]
+
+_THIS_DIR = Path(__file__).parent.resolve()
+_ROOT_DIR = _THIS_DIR.parent.parent.resolve()
+
+
+def _get_build(var, default=False):
+    if var not in os.environ:
+        return default
+
+    val = os.environ.get(var, "0")
+    trues = ["1", "true", "TRUE", "on", "ON", "yes", "YES"]
+    falses = ["0", "false", "FALSE", "off", "OFF", "no", "NO"]
+    if val in trues:
+        return True
+    if val not in falses:
+        print(f"WARNING: Unexpected environment variable value `{var}={val}`. " f"Expected one of {trues + falses}")
+    return False
+
+
+_BUILD_S3 = _get_build("BUILD_S3", False)
+
+
+def get_ext_modules():
+    if _BUILD_S3:
+        return [Extension(name="torchdata._torchdata", sources=[])]
+    else:
+        return []
+
+
+# Based off of pybiind cmake_example
+# https://github.com/pybind/cmake_example/blob/2440893c60ed2578fb127dc16e6b348fa0be92c1/setup.py
+# and torchaudio CMakeBuild()
+# https://github.com/pytorch/audio/blob/ece03edc3fc28a1ce2c28ef438d2898ed0a78d3f/tools/setup_helpers/extension.py#L65
+class CMakeBuild(build_ext):
+    def build_extension(self, ext):
+        # Because the following `cmake` command will build all of `ext_modules`` at the same time,
+        # we would like to prevent multiple calls to `cmake`.
+        # Therefore, we call `cmake` only for `torchdata._torchdata`,
+        # in case `ext_modules` contains more than one module.
+        if ext.name != "torchdata._torchdata":
+            return
+
+        extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+
+        # required for auto-detection of auxiliary "native" libs
+        if not extdir.endswith(os.path.sep):
+            extdir += os.path.sep
+
+        debug = int(os.environ.get("DEBUG", 0)) if self.debug is None else self.debug
+        cfg = "Debug" if debug else "Release"
+
+        cmake_args = [
+            f"-DCMAKE_BUILD_TYPE={cfg}",  # not used on MSVC, but no harm
+            f"-DCMAKE_PREFIX_PATH={torch.utils.cmake_prefix_path}",
+            f"-DCMAKE_INSTALL_PREFIX={extdir}",
+            f"-DCMAKE_LIBRARY_OUTPUT_DIRECTORY={extdir}",
+            f"-DPYTHON_EXECUTABLE={sys.executable}",
+            "-DCMAKE_CXX_FLAGS=-fPIC",
+            f"-DBUILD_S3:BOOL={'ON' if _BUILD_S3 else 'OFF'}",
+        ]
```

Comment:
Could we directly provide `Python_EXECUTABLE`, `Python_INCLUDE_DIRS`, `Python_LIBRARIES` here to bypass `find_package(python3)`

Reply:
Thanks. That's what I'm trying as well. Just pushed a commit, check if the paths are correct.

Reply:
I was also looking at this from pybind11, but somehow the `PYTHON_EXECUTABLE` wasn't used in the CMake. Maybe we don't even need the find_package(Python3) at all.

Reply:
Yeah. It's probably true that `PYTHON_EXECUTABLE` is not used as executable won't be used for compilation.